design/external-authorization-design.md

```diff
@@ -0,0 +1,710 @@
+
+If the `Service` refers to a Kubernetes `ExternalName`, Contour should program Envoy to send the traffic to the external destination.
+
+The `ExtensionService` CRD re-uses the `Service` type from the `projectcontour.io/v1` API.
+However, the setting following fields can generate a validation errors:
+
+- `Protocol` may only be set to `h2` or `h2c` (the default should be `h2`).
+
+Note that the `Service` type does not include a field for a namespace name.
+This constrains `ExtensionService` resources to be in the same namespace as the `Service` they expose,
+ensuring that whoever creates the `ExtensionService` also has authority over the `Service` backing it.
+
+### HTTPProxy Changes
+
+```Go
+// ExtensionReference names a Contour extension resource (ExtensionService
+// by default).
+type ExtensionExtensionReference struct {
+	// API version of the referent.
+        // If this field has no value, Contour will use a default of "projectcontour.io/v1alpha1".
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,5,opt,name=apiVersion"`
+	// Kind of the referent.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+        // If this field has no value, Contour will use a default of "ExtensionService".
+	// +optional
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+	// Namespace of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+        // If this field has no value, Contour will use the namespace of the enclosing HTTPProxy resource.
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +required
+	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+}
+
+type AuthorizationPolicy struct {
+	// When true, this field disables client request authentication
+	// for the scope of the policy.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value literal strings that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merged
+	// such that the inner scope overrides matching keys from the
+	// outer scope.
+	//
+	// +optional
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationExtension struct {
+        // ExtensionRef names the authorization service.
+	// +required
+	ExtensionRef ExtensionReference `json:"extensionRef"`
+
+        // AuthPolicy is the default authorization policy applied to client requests.
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
+
+        // If FailOpen is true, the client request is forwarded to the upstream service
```

Comment:
Is this controlling the `status_on_error` on Envoy API ([link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz))? 
Maybe document the default behavior here to clarify that HTTP 403 Forbidden will be returned by default if authorization server fails to respond.

Reply:
Yup, definitely worth make that clear in the docs.